### ISSUE TITLE
Relogin triggered on many ocassions

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#10440]
While testing the current state of `release` for a potential short term release, I stumbled upon a general slowness, e.g. when opening a cache, doing a nearby search, adding a cache to watchlist, and many others.

Looking at the debug log I can clearly see, that all of the above mentioned actions trigger a rerun of the login process in the background. 
Example scenario:
- Turn debug logging on
- Start c:geo, wait until logged in
- Perform nearby search

Result:
- The debug log shows the GET command for the search and the corresponding reply
- Triggered by this reply (?) the login process is started by c:geo
- The nearby search is done again 
(Luckily there is no endless loop created)

It seems our "still logged in check" is broken and assumes not being logged in and therefore applying a login.

Log of above mentioned example:
```
2021-04-22 19:57:19.561 20094-20491/cgeo.geocaching D/cgeo: [OkHttp] 200 (519 ms) GET https://www.geocaching.com/seek/nearest.aspx?lat=51.00205&lng=7.00535&tx=9a79e6ce-3344-409c-bbe9-496530baf758 (http/1.1)
2021-04-22 19:57:19.592 20094-20491/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/account/signin
2021-04-22 19:57:20.092 20094-20491/cgeo.geocaching D/cgeo: [OkHttp] 200 (500 ms) GET https://www.geocaching.com/account/signin (http/1.1) (=> https://www.geocaching.com/play/search)
2021-04-22 19:57:20.132 20094-20158/cgeo.geocaching I/cgeo: [network--4] Already logged in Geocaching.com as Zubu (BASIC)
2021-04-22 19:57:20.148 20094-20491/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/play/culture/set?model.SelectedCultureCode=en-US
2021-04-22 19:57:21.047 20094-20491/cgeo.geocaching D/cgeo: [OkHttp] 200 (899 ms) GET https://www.geocaching.com/play/culture/set?model.SelectedCultureCode=en-US (http/1.1) (=> https://www.geocaching.com/play/search)
2021-04-22 19:57:21.071 20094-20158/cgeo.geocaching I/cgeo: [network--4] changed language on geocaching.com to English
2021-04-22 19:57:21.099 20094-20491/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/account/signin
2021-04-22 19:57:21.601 20094-20491/cgeo.geocaching D/cgeo: [OkHttp] 200 (502 ms) GET https://www.geocaching.com/account/signin (http/1.1) (=> https://www.geocaching.com/play/search)
2021-04-22 19:57:21.638 20094-20126/cgeo.geocaching I/cgeo.geocachin: Explicit concurrent copying GC freed 90407(3283KB) AllocSpace objects, 20(4524KB) LOS objects, 70% free, 10057KB/33MB, paused 87us total 66.370ms
2021-04-22 19:57:21.642 20094-20158/cgeo.geocaching I/cgeo: [network--4] Already logged in Geocaching.com as Zubu (BASIC)
2021-04-22 19:57:21.658 20094-20491/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/play/culture/set?model.SelectedCultureCode=en-US
2021-04-22 19:57:22.549 20094-20491/cgeo.geocaching D/cgeo: [OkHttp] 200 (891 ms) GET https://www.geocaching.com/play/culture/set?model.SelectedCultureCode=en-US (http/1.1) (=> https://www.geocaching.com/play/search)
2021-04-22 19:57:22.573 20094-20158/cgeo.geocaching I/cgeo: [network--4] changed language on geocaching.com to English
2021-04-22 19:57:22.600 20094-20491/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/account/settings/homelocation
2021-04-22 19:57:22.603 20094-20620/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/seek/nearest.aspx?lat=51.00205&lng=7.00535&tx=9a79e6ce-3344-409c-bbe9-496530baf758
2021-04-22 19:57:22.840 20094-20491/cgeo.geocaching D/cgeo: [OkHttp] 200 (239 ms) GET https://www.geocaching.com/account/settings/homelocation (http/1.1)
2021-04-22 19:57:23.697 20094-20620/cgeo.geocaching D/cgeo: [OkHttp] 200 (1094 ms) GET https://www.geocaching.com/seek/nearest.aspx?lat=51.00205&lng=7.00535&tx=9a79e6ce-3344-409c-bbe9-496530baf758 (http/1.1)
```

**Comment by moving-bits** [comment]
Actually we do have two problems: 
- Login (or better: username) is not recognized on non-login pages (which triggers a second login), 
- and the language detection does not work any longer.

For part 1 I may have a solution, will prepare a PR for it, whereas part 2 needs some different solution as there is no language selector any more on the page.

**Comment by Lineflyer** [comment]
> [...] as there is no language selector any more on the page.

What page is that?

In that sense we should also check, whether date format detection works normal.

**Comment by moving-bits** [comment]
Found a new way for language detection, based on server parameters, will add that to the PR.

**Comment by Lineflyer** [comment]
@moving-bits 
Tested based on current HEAD of `release`.
Its still different to the working state before website changes:
On network operation there is still an unneeded GET to  https://www.geocaching.com/account/signin and some followup action

```
2021-04-22 23:02:35.276 31461-32234/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/seek/cache_details.aspx?decrypt=y&wp=GC4J4FK&log=y&numlogs=0
2021-04-22 23:02:36.671 31461-32234/cgeo.geocaching D/cgeo: [OkHttp] 200 (1395 ms) GET https://www.geocaching.com/seek/cache_details.aspx?decrypt=y&wp=GC4J4FK&log=y&numlogs=0 (http/1.1) (=> https://www.geocaching.com/geocache/GC4J4FK_tb-hotel-d162?decrypt=y&log=y&numlogs=0)
2021-04-22 23:02:37.291 31461-32234/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/account/signin
2021-04-22 23:02:37.814 31461-32234/cgeo.geocaching D/cgeo: [OkHttp] 200 (523 ms) GET https://www.geocaching.com/account/signin (http/1.1) (=> https://www.geocaching.com/play/search)
2021-04-22 23:02:38.226 31461-31924/cgeo.geocaching I/cgeo: [refresh--2] Already logged in Geocaching.com as Zubu (BASIC)
2021-04-22 23:02:38.226 31461-31924/cgeo.geocaching I/cgeo: [refresh--2] Geocaching.com language already set to English
2021-04-22 23:02:38.230 31461-32234/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/account/settings/homelocation
2021-04-22 23:02:38.232 31461-32242/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/seek/cache_details.aspx?decrypt=y&wp=GC4J4FK&log=y&numlogs=0
2021-04-22 23:02:38.477 31461-32234/cgeo.geocaching D/cgeo: [OkHttp] 200 (247 ms) GET https://www.geocaching.com/account/settings/homelocation (http/1.1)
2021-04-22 23:02:39.564 31461-32242/cgeo.geocaching D/cgeo: [OkHttp] 200 (1332 ms) GET https://www.geocaching.com/seek/cache_details.aspx?decrypt=y&wp=GC4J4FK&log=y&numlogs=0 (http/1.1) (=> https://www.geocaching.com/geocache/GC4J4FK_tb-hotel-d162?decrypt=y&log=y&numlogs=0)
2021-04-22 23:02:40.219 31461-32242/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/seek/geocache.logbook?tkn=4OB3GFHLRR3CGAUGOYCJ5F5VO3SCW3UC6PXUL53I56QFEL5YQ4ERXC2DWKMNAIYBO4OPWN7WIT65PBTZG3LVFO6JWZZQODVFXQNB2AY2ODVQEB7HIXDCYH56H4DR252TEAFQFH7B6EG332A26KSKYCYE4CVVBNNVRKVSKQHS6L3FPJ4SZAVDBEGK2RD6PZ5AW2TMUQ4Q7GHXODDB2CKBL4NM4X7CYNQD6O2L4NWCMRUAU74BOPP4DUSIRDSALO4HMM6UA5NCMU7LY66AFO66ONT5TDNYFWXMD7BF4B4KMNGY5YZYOVTGDGWIJILPFZTGR7PYGKVZWPQYC3ZPKFMLZZO7BE&idx=1&num=35&decrypt=false
2021-04-22 23:02:40.219 31461-32234/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/seek/geocache.logbook?tkn=4OB3GFHLRR3CGAUGOYCJ5F5VO3SCW3UC6PXUL53I56QFEL5YQ4ERXC2DWKMNAIYBO4OPWN7WIT65PBTZG3LVFO6JWZZQODVFXQNB2AY2ODVQEB7HIXDCYH56H4DR252TEAFQFH7B6EG332A26KSKYCYE4CVVBNNVRKVSKQHS6L3FPJ4SZAVDBEGK2RD6PZ5AW2TMUQ4Q7GHXODDB2CKBL4NM4X7CYNQD6O2L4NWCMRUAU74BOPP4DUSIRDSALO4HMM6UA5NCMU7LY66AFO66ONT5TDNYFWXMD7BF4B4KMNGY5YZYOVTGDGWIJILPFZTGR7PYGKVZWPQYC3ZPKFMLZZO7BE&idx=1&num=35&decrypt=false&sf=true
2021-04-22 23:02:40.220 31461-32246/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/seek/geocache.logbook?tkn=4OB3GFHLRR3CGAUGOYCJ5F5VO3SCW3UC6PXUL53I56QFEL5YQ4ERXC2DWKMNAIYBO4OPWN7WIT65PBTZG3LVFO6JWZZQODVFXQNB2AY2ODVQEB7HIXDCYH56H4DR252TEAFQFH7B6EG332A26KSKYCYE4CVVBNNVRKVSKQHS6L3FPJ4SZAVDBEGK2RD6PZ5AW2TMUQ4Q7GHXODDB2CKBL4NM4X7CYNQD6O2L4NWCMRUAU74BOPP4DUSIRDSALO4HMM6UA5NCMU7LY66AFO66ONT5TDNYFWXMD7BF4B4KMNGY5YZYOVTGDGWIJILPFZTGR7PYGKVZWPQYC3ZPKFMLZZO7BE&idx=1&num=35&decrypt=false&sp=true
2021-04-22 23:02:40.399 31461-32246/cgeo.geocaching D/cgeo: [OkHttp] 200 (179 ms) GET https://www.geocaching.com/seek/geocache.logbook?tkn=4OB3GFHLRR3CGAUGOYCJ5F5VO3SCW3UC6PXUL53I56QFEL5YQ4ERXC2DWKMNAIYBO4OPWN7WIT65PBTZG3LVFO6JWZZQODVFXQNB2AY2ODVQEB7HIXDCYH56H4DR252TEAFQFH7B6EG332A26KSKYCYE4CVVBNNVRKVSKQHS6L3FPJ4SZAVDBEGK2RD6PZ5AW2TMUQ4Q7GHXODDB2CKBL4NM4X7CYNQD6O2L4NWCMRUAU74BOPP4DUSIRDSALO4HMM6UA5NCMU7LY66AFO66ONT5TDNYFWXMD7BF4B4KMNGY5YZYOVTGDGWIJILPFZTGR7PYGKVZWPQYC3ZPKFMLZZO7BE&idx=1&num=35&decrypt=false&sp=true (http/1.1)
2021-04-22 23:02:40.463 31461-32242/cgeo.geocaching D/cgeo: [OkHttp] 200 (244 ms) GET https://www.geocaching.com/seek/geocache.logbook?tkn=4OB3GFHLRR3CGAUGOYCJ5F5VO3SCW3UC6PXUL53I56QFEL5YQ4ERXC2DWKMNAIYBO4OPWN7WIT65PBTZG3LVFO6JWZZQODVFXQNB2AY2ODVQEB7HIXDCYH56H4DR252TEAFQFH7B6EG332A26KSKYCYE4CVVBNNVRKVSKQHS6L3FPJ4SZAVDBEGK2RD6PZ5AW2TMUQ4Q7GHXODDB2CKBL4NM4X7CYNQD6O2L4NWCMRUAU74BOPP4DUSIRDSALO4HMM6UA5NCMU7LY66AFO66ONT5TDNYFWXMD7BF4B4KMNGY5YZYOVTGDGWIJILPFZTGR7PYGKVZWPQYC3ZPKFMLZZO7BE&idx=1&num=35&decrypt=false (http/1.1)
2021-04-22 23:02:40.961 31461-32234/cgeo.geocaching D/cgeo: [OkHttp] 200 (742 ms) GET https://www.geocaching.com/seek/geocache.logbook?tkn=4OB3GFHLRR3CGAUGOYCJ5F5VO3SCW3UC6PXUL53I56QFEL5YQ4ERXC2DWKMNAIYBO4OPWN7WIT65PBTZG3LVFO6JWZZQODVFXQNB2AY2ODVQEB7HIXDCYH56H4DR252TEAFQFH7B6EG332A26KSKYCYE4CVVBNNVRKVSKQHS6L3FPJ4SZAVDBEGK2RD6PZ5AW2TMUQ4Q7GHXODDB2CKBL4NM4X7CYNQD6O2L4NWCMRUAU74BOPP4DUSIRDSALO4HMM6UA5NCMU7LY66AFO66ONT5TDNYFWXMD7BF4B4KMNGY5YZYOVTGDGWIJILPFZTGR7PYGKVZWPQYC3ZPKFMLZZO7BE&idx=1&num=35&decrypt=false&sf=true (http/1.1)
```

IMHO the initial call to .../signin should not even happen. 
"Only" the unneeded language change was solved so far.

**Comment by Lineflyer** [comment]
@cgeo/team If anyone has more knowledge or even was involved in those code parts, feel free to help. For @moving-bits and myself its more or less reverse engineering at this point and we don't want to make it worse.

**Comment by Lineflyer** [comment]
AFAIR we used some flag to store whether we assume, that the user is still logged in (most likely checking if the user name is present on the last parsed page).
If this flag is set to FALSE the next network interaction (where the flag is checked) will trigger a relogin before executing the network request.

If my assumption above is correct, than this flag seems to return false now all the time and this needs to be fixed to prevent triggering the unneeded login.

Fixed until now is the unneeded loop for language switch and the login procedure itself.

**Comment by Lineflyer** [comment]
I crawled through some source texts of the pages on geocaching.com. To me it looks quite simple to detect being logged in. There is a variable "username" in `Window['headerSettings`']` which either is followed by the username itself or contains `null` on all pages I checked.

**Comment by baiti** [comment]
Can't look at the source right now but isn't it the `gspkauth` cookie that indicates an authenticated session? Unfortunately the gspkauth is not a JWT so you can't decode it but the first try should always be to present this cookie and if it is still valid, we're in, no login required. If not, we need to log in again and then persist gspkauth. It lasts weeks rather than days. as far as I remember.

**Comment by baiti** [comment]
SInce I had approx 30min left I scanned some code, specifically GCLogin.java and I haven't seen any code that deals with cookies.  The only variable that seems of interes is "requestVerificationToken". I certainly don't know the history of the code, nor do I know if and how cookies are handled by the http(s) library. I can only tell, from my own experiments, `gspkauth` is the one and only cookie needed to make authenticated requests. Of course this cookie can't be parsed ot of the web page itsself, it comes down in the response headers. The http library should provide a way to manage what is sometimes called a "cookie jar" and it should allow to put these cookies back into the request header of subsequent calls. Can't investigate any deeper at this time. Perhaps later in the afternoon.

**Comment by moving-bits** [comment]
> I crawled through some source texts of the pages on geocaching.com. To me it looks quite simple to detect being logged in. There is a variable "username" in `Window['headerSettings']` which either is followed by the username itself or contains `null` on all pages I checked.

With #10436 we had switched from the old approach to something similar, there was a `dataLayer` var on each page, containing the username as well. But as I've seen today, that var already got removed again from the source and extracted to a separate script call, so we have to adapt again (e. g. to `headerSettings`). We should probably use this to be back into the game before the weekend.

But if there is a more generalized way like the one @baiti describes, this could be a more sustainable solution in the future.

**Comment by baiti** [comment]
I looked a bit more at the code. There is no mention of a cookie or cookieJar anywhere. That leads me to believe that `isLoggedIn()` is determined with other means. It must be scraped out of the http body received and parsed via regex. I am sure this can be fxed such that it is working again but I am deeply convinced that a much better design would be to go for the gspkauth cookie. This would require some refactoring though.  Should we ever consider to go via the partner API my suggestion would be not to invest into optimizing what then would be "a dead horse". Should we however decide to stay away from that API I am convinced the GC Connector could be improved. A login should only be required at a frequency of perhaps once per month.

@moving-bits , you have touched the code recently, what are your thoughts with respect to gspkauth?

**Comment by Lineflyer** [comment]
Any chance on a shortterm fix?
Support mailbox and Google Plaz rating mail are eploding. Long time we did not have such an outage...our users seem no longer used to it...or there are even more people caching out there as last years (maybe du to Corona).

**Comment by bekuno** [comment]
I tested a bit around with the nightly. For me at least the main functions where fine and with good performance.
If no short term optimization is possible then we should release the nightly as an hot fix release.

**Comment by Lineflyer** [comment]
That would be the last resort.
I am a bit reluctant though, as the login page will be called quite often by all users then.

I wish I could help...I am quite convinced, that it cannot be so hard to solve with a matching regexp.

**Comment by moving-bits** [comment]
I've adapted the rexexp to their current source code and can re-authenticate with that. PR will follow. Still need to debug into #10440, though.

BTW: They have changed their source twice since our fix from Wednesday...

**Comment by Lineflyer** [comment]
@moving-bits 
Does this need change:
ˋˋˋ
// Info box top-right

public static final Pattern PATTERN_LOGIN_NAME = Pattern.compile("class=\"user-name\"[^>]*>(.*?)</span>", Pattern.DOTALL);
ˋˋˋ

**Comment by moving-bits** [comment]
> Does this need change:
> `public static final Pattern PATTERN_LOGIN_NAME = Pattern.compile("class="user-name"[^>]_>(._?)", Pattern.DOTALL);`

From the source: this is used for the tests only. So needs to be updated, but should not trigger a relogin.

**Comment by moving-bits** [comment]
Using the version from #10450 I've tried the "relogin" scenario described above. Seems to be ok here from what I read from the log (I've filtered out all non-okhttp entries):
```
2021-04-23 23:15:57.673 28199-28910/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/seek/nearest.aspx?lat=40.712775&lng=-74.005972&tx=9a79e6ce-3344-409c-bbe9-496530baf758
2021-04-23 23:15:58.049 28199-28910/cgeo.geocaching D/cgeo: [OkHttp] 200 (375 ms) GET https://www.geocaching.com/seek/nearest.aspx?lat=40.712775&lng=-74.005972&tx=9a79e6ce-3344-409c-bbe9-496530baf758 (http/1.1)
2021-04-23 23:16:42.757 28199-28910/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/images/icons/compass/NW.gif
2021-04-23 23:16:42.791 28199-28921/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/images/icons/compass/W.gif
2021-04-23 23:16:42.844 28199-28923/cgeo.geocaching D/cgeo: [OkHttp] GET https://www.geocaching.com/images/icons/compass/E.gif
2021-04-23 23:16:42.933 28199-28910/cgeo.geocaching D/cgeo: [OkHttp] 200 (175 ms) GET https://www.geocaching.com/images/icons/compass/NW.gif (http/1.1)
2021-04-23 23:16:43.554 28199-28923/cgeo.geocaching D/cgeo: [OkHttp] 200 (710 ms) GET https://www.geocaching.com/images/icons/compass/E.gif (http/1.1)
2021-04-23 23:16:43.563 28199-28921/cgeo.geocaching D/cgeo: [OkHttp] 200 (772 ms) GET https://www.geocaching.com/images/icons/compass/W.gif (http/1.1)
2021-04-23 23:16:44.768 28199-28921/cgeo.geocaching D/cgeo: [OkHttp] POST https://www.geocaching.com/seek/nearest.aspx?lat=40.712775&lng=-74.005972&tx=9a79e6ce-3344-409c-bbe9-496530baf758
2021-04-23 23:16:45.193 28199-28921/cgeo.geocaching D/cgeo: [OkHttp] 200 (425 ms) POST https://www.geocaching.com/seek/nearest.aspx?lat=40.712775&lng=-74.005972&tx=9a79e6ce-3344-409c-bbe9-496530baf758 (http/1.1)
```

(Test contained performing a nearby search and pressing "show more results" once.)

**Comment by Lineflyer** [comment]
This looks totally solved by #10450
